ni_measurementlink_service/_internal/parameter/serialization_strategy.py

```diff
@@ -1,257 +1,212 @@
 """Serialization Strategy."""
+from __future__ import annotations
 
-from typing import Any, Callable
+import sys
+import typing
+from typing import Any, Callable, Dict, TypeVar, cast
 
 from google.protobuf import type_pb2
-from google.protobuf.internal import decoder
-from google.protobuf.internal import encoder
+from google.protobuf.internal import decoder, encoder
+from google.protobuf.message import Message
 
+if typing.TYPE_CHECKING:
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
 
-def _scalar_encoder(encoder) -> Callable[[int], Callable]:
-    """Abstract Specific Encoder(Callable) as Scalar Encoder Callable that takes in field index.
+Key = TypeVar("Key")
 
-    Args
-    ----
-       encoder (Callable[[int, bool, bool], Callable]): Specific encoder that takes in
-       field_index, is_repeated, is_packed and returns the Low-level Encode Callable.
+WriteFunction: TypeAlias = Callable[[bytes], int]
+Encoder: TypeAlias = Callable[[WriteFunction, bytes, bool], int]
+PartialEncoderConstructor: TypeAlias = Callable[[int], Encoder]
+EncoderConstructor: TypeAlias = Callable[[int, bool, bool], Encoder]
 
-    Returns
-    -------
-        Callable[[int],Callable]: Callable Encoder for scalar types that takes
-        in field_index and returns the Low-level Encode Callable.
+Decoder: TypeAlias = Callable[[memoryview, int, int, Message, Dict[Key, Any]], int]
+PartialDecoderConstructor: TypeAlias = Callable[[int, Key], Decoder]
+NewDefault: TypeAlias = Callable[[Message], Message]
+DecoderConstructor: TypeAlias = Callable[[int, bool, bool, Key, NewDefault], Decoder]
 
+
+def _scalar_encoder(encoder: EncoderConstructor) -> PartialEncoderConstructor:
+    """Constructs a scalar encoder constructor.
+
+    Takes a field index and returns an Encoder.
+
+    This class returns the Encoder with is_repeated set to False
+    and is_packed set to False.
     """
 
-    def scalar_encoder(field_index):
+    def scalar_encoder(field_index: int) -> Encoder:
         is_repeated = False
         is_packed = False
         return encoder(field_index, is_repeated, is_packed)
 
     return scalar_encoder
 
 
-def _vector_encoder(encoder, is_packed=True) -> Callable[[int], Callable]:
-    """Abstract Specific Encoder(Callable) as Vector Encoder Callable that takes in field index.
-
-    Args
-    ----
-       encoder (Callable[[int, bool, bool], Callable]): Specific encoder(Callable) that takes in
-       field_index, is_repeated, is_packed and returns the Low-level Encode Callable.
+def _vector_encoder(encoder: EncoderConstructor, is_packed=True) -> PartialEncoderConstructor:
```

Comment:
mypy doesn't seem to infer types based on default arguments. We probably get this wrong a lot, but at least annotate is_repeated and is_packed everywhere.

```suggestion
def _vector_encoder(encoder: EncoderConstructor, is_packed: bool = True) -> PartialEncoderConstructor:
```